SpatialGDK/Source/SpatialGDKTests/SpatialGDK/LoadBalancing/ReferenceCountedLockingPolicy/ReferenceCountedLockingPolicyTest.cpp

```diff
@@ -0,0 +1,117 @@
+{
+	UWorld* TestWorld;
+	TMap<FName, AActor*> TestActors;
+	UReferenceCountedLockingPolicy* LockingPolicy;
+};
+
+struct TestDataDeleter
+{
+	void operator()(TestData* Data) const noexcept
+	{
+		Data->LockingPolicy->RemoveFromRoot();
+		delete Data;
+	}
+};
+
+TSharedPtr<TestData> MakeNewTestData()
+{
+	TSharedPtr<TestData> Data(new TestData, TestDataDeleter());
+	Data->LockingPolicy = NewObject<UReferenceCountedLockingPolicy>();
+	Data->LockingPolicy->AddToRoot();
+	return Data;
+}
+
+// Copied from AutomationCommon::GetAnyGameWorld()
+UWorld* GetAnyGameWorld()
+{
+	UWorld* World = nullptr;
+	const TIndirectArray<FWorldContext>& WorldContexts = GEngine->GetWorldContexts();
+	for (const FWorldContext& Context : WorldContexts)
+	{
+		if ((Context.WorldType == EWorldType::PIE || Context.WorldType == EWorldType::Game)
+			&& (Context.World() != nullptr))
+		{
+			World = Context.World();
+			break;
+		}
+	}
+
+	return World;
+}
+
+DEFINE_LATENT_AUTOMATION_COMMAND_ONE_PARAMETER(FWaitForWorld, TSharedPtr<TestData>, Data);
+bool FWaitForWorld::Update()
+{
+	Data->TestWorld = GetAnyGameWorld();
+
+	if (Data->TestWorld && Data->TestWorld->AreActorsInitialized())
+	{
+		AGameStateBase* GameState = Data->TestWorld->GetGameState();
+		if (GameState && GameState->HasMatchStarted())
+		{
+			return true;
+		}
+	}
+
+	return false;
+}
+
+DEFINE_LATENT_AUTOMATION_COMMAND_TWO_PARAMETER(FSpawnActor, TSharedPtr<TestData>, Data, FName, Handle);
+bool FSpawnActor::Update()
+{
+	FActorSpawnParameters SpawnParams;
+	SpawnParams.bNoFail = true;
+	SpawnParams.SpawnCollisionHandlingOverride = ESpawnActorCollisionHandlingMethod::AlwaysSpawn;
+
+	AActor* Actor = Data->TestWorld->SpawnActor<ADefaultPawn>(SpawnParams);
+	Data->TestActors.Add(Handle, Actor);
+
+	return true;
+}
+
+DEFINE_LATENT_AUTOMATION_COMMAND_TWO_PARAMETER(FWaitForActor, TSharedPtr<TestData>, Data, FName, Handle);
+bool FWaitForActor::Update()
+{
+	AActor* Actor = Data->TestActors[Handle];
+	return (IsValid(Actor) && Actor->IsActorInitialized() && Actor->HasActorBegunPlay());
+}
+
+DEFINE_LATENT_AUTOMATION_COMMAND_FOUR_PARAMETER(FTestIsLocked, FAutomationTestBase*, Test, TSharedPtr<TestData>, Data, FName, Handle, bool, bExpected);
+bool FTestIsLocked::Update()
+{
+	AActor* Actor = Data->TestActors[Handle];
+	const bool bIsLocked = Data->LockingPolicy->IsLocked(Actor);
+	Test->TestEqual(FString::Printf(TEXT("Is locked. Actual: %d. Expected: %d"), bIsLocked, bExpected), bIsLocked, bExpected);
+	return true;
+}
+
+}  // anonymous namespace
+
+
+REFERENCECOUNTEDLOCKINGPOLICY_TEST(GIVEN_an_actor_has_not_been_locked_WHEN_IsLocked_is_called_THEN_returns_false)
+{
+	AutomationOpenMap("/Engine/Maps/Entry");
+
+	TSharedPtr<TestData> Data = MakeNewTestData();
+	ADD_LATENT_AUTOMATION_COMMAND(FWaitForWorld(Data));
+	ADD_LATENT_AUTOMATION_COMMAND(FSpawnActor(Data, "Actor"));
+	ADD_LATENT_AUTOMATION_COMMAND(FWaitForActor(Data, "Actor"));
+	ADD_LATENT_AUTOMATION_COMMAND(FTestIsLocked(this, Data, "Actor", false));
```

Comment:
It would be useful to also have tests that use locking and verify it works as expected.

Reply:
If I'm not missing anything, there's currently one tests, that checks that the Actor is not locked by default?

Reply:
Calling AcquireLock or CanAcquireLock requires checking authority and authority intent which require the NetDriver which we have no way of mocking. Ultimately, better unit tests are blocked on fixing the authority model with migration.

Reply:
I see, since `CanAcqureLock` uses `NetDriver`, `PacakgeMap` and `StaticComponentView`, testing the Locking Policy gets really complicated.

Reply:
Yes. It's not meant to use any of them but Actor->HasAuthority currently returns the wrong value and needs to be fixed.

Reply:
What if for testing purposes we derive a class from `UReferenceCountedLockingPolicy` and only override `CanAcquireLock` function? Then we'll be able to test all the other functions.

Reply:
I like this idea as a short-term solution

Reply:
I think that current test is not enough. Could you add more tests to cover other `UReferenceCountedLockingPolicy` functions (potentially by overriding `CanAcquireLock` function)? Or create a follow-up task to write tests for the `UReferenceCountedLockingPolicy`?

Reply:
This is one of the rare cases where I'd rather have a follow up task (this was never meant to be the full suite of tests for this class). I think for us to have more tests now we either need to write slightly dodgy tests or make the implementation a lot worse.

Reply:
Notably because we now think that CanAcquireLock shouldn't be virtual at all.

Reply:
If we want to talk about this more it might be slightly quicker offline. But I do agree this one test is not really enough to mean anything.

Reply:
Please create a jira ticket for the follow-up task and resolve this conversation.

Reply:
https://improbableio.atlassian.net/browse/UNR-2352